File: src/html.rs
//! HTML assembly: combines Mermaid graphs, topology description, and optional
//! plan markdown into a self-contained HTML page using Askama templates.
//!
//! CSS and JS are embedded at build time via `include_str!()`. Plan markdown
//! is JSON-encoded and rendered client-side with marked + DOMPurify.

use askama::Template;

use crate::types::{Agent, ExecutionMode, Topology};

const CSS: &str = include_str!("../static/style.css");
const JS: &str = include_str!("../static/script.js");

#[derive(Template)]
#[template(path = "page.html", escape = "none")]
struct PageTemplate<'a> {
    task_summary: String,
    mode_label: String,
    mermaid_graphs: &'a [String],
    description: String,
    has_plan: bool,
    plan_markdown_json: String,
    phase_labels: Vec<String>,
    css: &'a str,
    js: &'a str,
}

/// Generates a self-contained HTML page from topology rendering outputs.
///
/// - `topology`: the parsed topology (used for task_summary, execution_mode)
/// - `mermaid_graphs`: output of `mermaid::generate()` — one graph per phase
/// - `description`: output of `describe::generate()` — human-readable overview
/// - `plan_markdown`: optional plan markdown content (from `--plan <file>`)
pub fn generate(
    topology: &Topology,
    mermaid_graphs: &[String],
    description: &str,
    plan_markdown: Option<&str>,
) -> String {
    let has_plan = plan_markdown.is_some();
    let plan_markdown_json = plan_markdown
        .map(|md| serde_json::to_string(md).expect("JSON string encoding"))
        .unwrap_or_default();

    let template = PageTemplate {
        task_summary: html_escape(&topology.task_summary),
        mode_label: build_mode_label(topology),
        mermaid_graphs,
        description: html_escape(description),
        has_plan,
        plan_markdown_json,
        phase_labels: build_phase_labels(topology, mermaid_graphs.len()),
        css: CSS,
        js: JS,
    };

    template.render().expect("template rendering failed")
}

pub fn html_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
}

fn build_mode_label(topology: &Topology) -> String {
    let top = match topology.execution_mode {
        ExecutionMode::Team => "team",
        ExecutionMode::Subagents => "subagents",
    };

    let has_nested_team = has_nested_mode(&topology.agents, &ExecutionMode::Team);
    let has_nested_subagents = has_nested_mode(&topology.agents, &ExecutionMode::Subagents);

    match (&topology.execution_mode, has_nested_team, has_nested_subagents) {
        (ExecutionMode::Subagents, true, _) => "subagents + team".to_string(),
        (ExecutionMode::Team, _, true) => "team + subagents".to_string(),
        _ => top.to_string(),
    }
}

fn has_nested_mode(agents: &[Agent], target: &ExecutionMode) -> bool {
    agents.iter().any(|a| {
        if a.execution_mode.as_ref() == Some(target) {
            return true;
        }
        if let Some(children) = &a.agents {
            return has_nested_mode(children, target);
        }
        false
    })
}

fn build_phase_labels(topology: &Topology, graph_count: usize) -> Vec<String> {
    if topology.execution_mode == ExecutionMode::Subagents && graph_count > 1 {
        (1..=graph_count)
            .map(|i| format!("Phase {}", i))
            .collect()
    } else {
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{Model, Output};

    fn make_agent(id: &str, model: Model, blocked_by: Vec<&str>) -> Agent {
        Agent {
            id: id.into(),
            role: "does stuff".into(),
            model,
            tools: vec!["Read".into()],
            blocked_by: blocked_by.into_iter().map(String::from).collect(),
            background: false,
            output: Output::Inline,
            produces: None,
            execution_mode: None,
            agents: None,
        }
    }

    fn make_topology(mode: ExecutionMode, agents: Vec<Agent>) -> Topology {
        Topology {
            task_summary: "Build a widget".into(),
            execution_mode: mode,
            agents,
        }
    }

    fn minimal_html() -> String {
        let t = make_topology(
            ExecutionMode::Subagents,
            vec![make_agent("builder", Model::Sonnet, vec![])],
        );
        let graphs = vec!["graph TD\n    main-->builder".into()];
        generate(&t, &graphs, "1. builder (sonnet)", None)
    }

    #[test]
    fn contains_title() {
        let html = minimal_html();
        assert!(html.contains("<title>Topology: Build a widget</title>"));
    }

    #[test]
    fn title_escapes_html() {
        let t = Topology {
            task_summary: "<script>alert(1)</script>".into(),
            execution_mode: ExecutionMode::Subagents,
            agents: vec![make_agent("a", Model::Haiku, vec![])],
        };
        let html = generate(&t, &["graph TD".into()], "desc", None);
        assert!(html.contains("&lt;script&gt;"));
        assert!(!html.contains("<script>alert"));
    }

    #[test]
    fn contains_mermaid_cdn() {
        let html = minimal_html();
        assert!(html.contains("mermaid@11.12.2"));
    }

    #[test]
    fn no_plan_single_column() {
        let html = minimal_html();
        assert!(html.contains("class=\"no-plan\""));
        assert!(!html.contains("<aside class=\"plan-panel\">"));
        assert!(!html.contains("marked@"));
        assert!(!html.contains("dompurify@"));
    }

    #[test]
    fn with_plan_two_column() {
        let t = make_topology(
            ExecutionMode::Subagents,
            vec![make_agent("a", Model::Sonnet, vec![])],
        );
        let html = generate(
            &t,
            &["graph TD".into()],
            "desc",
            Some("# My Plan\n- step 1"),
        );
        assert!(html.contains("class=\"has-plan\""));
        assert!(html.contains("plan-panel"));
        assert!(html.contains("marked@15.0.7"));
        assert!(html.contains("dompurify@3.2.4"));
        assert!(html.contains("window.__planMarkdown"));
    }

    #[test]
    fn plan_markdown_json_encoded() {
        let t = make_topology(
            ExecutionMode::Subagents,
            vec![make_agent("a", Model::Sonnet, vec![])],
        );
        let html = generate(
            &t,
            &["graph TD".into()],
            "desc",
            Some("test \"quotes\" and\nnewlines"),
        );
        // serde_json::to_string wraps in quotes and escapes
        assert!(html.contains(r#"\"quotes\""#));
        assert!(html.contains(r#"\n"#));
    }

    #[test]
    fn phase_labels_multi_step() {
        let t = make_topology(
            ExecutionMode::Subagents,
            vec![
                make_agent("a", Model::Sonnet, vec![]),
                make_agent("b", Model::Sonnet, vec!["a"]),
            ],
        );
        let graphs = vec!["graph TD\n    a".into(), "graph TD\n    b".into()];
        let html = generate(&t, &graphs, "desc", None);
        assert!(html.contains("Phase 1"));
        assert!(html.contains("Phase 2"));
    }

    #[test]
    fn no_phase_labels_single_step() {
        let html = minimal_html();
        assert!(!html.contains("Phase 1"));
    }

    #[test]
    fn no_phase_labels_team() {
        let t = make_topology(
            ExecutionMode::Team,
            vec![
                make_agent("a", Model::Sonnet, vec![]),
                make_agent("b", Model::Sonnet, vec!["a"]),
            ],
        );
        let graphs = vec!["graph TD".into()];
        let html = generate(&t, &graphs, "desc", None);
        assert!(!html.contains("<h3 class=\"phase-label\">"));
    }

    #[test]
    fn mode_label_combined() {
        let mut parent = make_agent("parent", Model::Opus, vec![]);
        parent.execution_mode = Some(ExecutionMode::Team);
        parent.agents = Some(vec![make_agent("child", Model::Haiku, vec![])]);
        let t = make_topology(ExecutionMode::Subagents, vec![parent]);
        let html = generate(&t, &["graph TD".into()], "desc", None);
        assert!(html.contains("subagents + team"));
    }

    #[test]
    fn mode_label_simple() {
        let html = minimal_html();
        assert!(html.contains("Mode: subagents"));
        assert!(!html.contains("+ team"));
    }

    #[test]
    fn contains_legend() {
        let html = minimal_html();
        assert!(html.contains("Legend"));
        assert!(html.contains("swatch-haiku"));
        assert!(html.contains("swatch-sonnet"));
        assert!(html.contains("swatch-opus"));
        assert!(html.contains("swatch-main"));
        assert!(html.contains("shape-rect"));
        assert!(html.contains("shape-pill"));
    }

    #[test]
    fn embeds_css_js() {
        let html = minimal_html();
        assert!(html.contains("planview styles"));
        assert!(html.contains("planview client-side JS"));
    }

    #[test]
    fn mermaid_graphs_embedded() {
        let t = make_topology(
            ExecutionMode::Subagents,
            vec![make_agent("a", Model::Sonnet, vec![])],
        );
        let graphs = vec!["graph TD\n    main-->builder".into()];
        let html = generate(&t, &graphs, "desc", None);
        assert!(html.contains("graph TD\n    main-->builder"));
    }

    #[test]
    fn description_embedded() {
        let t = make_topology(
            ExecutionMode::Subagents,
            vec![make_agent("a", Model::Sonnet, vec![])],
        );
        let html = generate(&t, &["graph TD".into()], "1. builder (sonnet)\n  tools: Read", None);
        assert!(html.contains("1. builder (sonnet)"));
    }

    #[test]
    fn fixture_minimal_full_pipeline() {
        let json = std::fs::read_to_string("tests/fixtures/valid_minimal.json").unwrap();
        let t: Topology = serde_json::from_str(&json).unwrap();
        let graphs = crate::mermaid::generate(&t);
        let desc = crate::describe::generate(&t);
        let html = generate(&t, &graphs, &desc, None);
        assert!(html.contains("<!DOCTYPE html>"));
        assert!(html.contains("</html>"));
        assert!(html.contains(&t.task_summary));
    }

    #[test]
    fn fixture_nested_with_plan() {
        let json = std::fs::read_to_string("tests/fixtures/valid_nested.json").unwrap();
        let t: Topology = serde_json::from_str(&json).unwrap();
        let graphs = crate::mermaid::generate(&t);
        let desc = crate::describe::generate(&t);
        let html = generate(&t, &graphs, &desc, Some("# Test Plan\n\nSome content"));
        assert!(html.contains("class=\"has-plan\""));
        assert!(html.contains("window.__planMarkdown"));
    }
}

File: ts/html.ts
import { Eta } from "eta";
import { CSS, JS, PAGE_TEMPLATE, PLAN_TEMPLATE } from "./assets.generated.js";
import { mermaid } from "./mermaid.js";
import { buildOverviewMd, unitIdPrefix } from "./render-md.js";
import type { Agent, ExecutionMode, Plan, Topology } from "./types.js";

const eta = new Eta({ autoEscape: false });

export function htmlEscape(s: string): string {
  return s
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
}

/**
 * JSON-encodes a string for embedding inside a `<script>` block. Beyond
 * normal JSON escaping, replaces `</` with `<\/` so the HTML tokenizer can't
 * be tricked into closing the script tag mid-payload, and escapes `<!--` for
 * older browsers.
 */
function jsonForScript(s: string): string {
  return JSON.stringify(s).replaceAll("</", "<\\/").replaceAll("<!--", "<\\!--");
}

function jsonArrayForScript(arr: string[]): string {
  return JSON.stringify(arr)
    .replaceAll("</", "<\\/")
    .replaceAll("<!--", "<\\!--");
}

export function renderTopologyHtml(
  topology: Topology,
  mermaidGraphs: string[],
  description: string,
  planMarkdown?: string,
): string {
  const hasPlan = planMarkdown !== undefined;
  const locals = {
    task_summary: htmlEscape(topology.task_summary),
    mode_label: buildModeLabel(topology),
    mermaid_graphs: mermaidGraphs,
    description: htmlEscape(description),
    has_plan: hasPlan,
    plan_markdown_json: hasPlan ? jsonForScript(planMarkdown) : "",
    phase_labels: buildPhaseLabels(topology, mermaidGraphs.length),
    css: CSS,
    js: JS,
  };
  const result = eta.renderString(PAGE_TEMPLATE, locals);
  return result;
}

interface UnitCard {
  index: number;
  anchor: string;
  prefix: string;
  title: string;
  summary: string;
  blocked_by_label: string;
  agents_label: string;
  has_topology: boolean;
  mermaid_graphs: string[];
  review_steps: string[];
}

export function renderPlanHtml(plan: Plan, dirName: string): string {
  const overviewMd = buildOverviewMd(plan, dirName);
  const unitBodies = plan.units.map((u) => u.body_markdown);
  const topologyCount = plan.units.filter((u) => u.topology !== undefined)
    .length;

  const units: UnitCard[] = plan.units.map((unit, i) => {
    const prefix = unitIdPrefix(unit.id) ?? unit.id;
    const blockedByLabel =
      unit.blocked_by.length === 0
        ? "—"
        : unit.blocked_by.map(htmlEscape).join(", ");
    const agentsLabel =
      unit.agents_involved && unit.agents_involved.length > 0
        ? unit.agents_involved.map(htmlEscape).join(", ")
        : "main only";
    const mermaidGraphs = unit.topology ? mermaid(unit.topology) : [];
    return {
      index: i,
      anchor: `unit-${htmlEscape(unit.id)}`,
      prefix,
      title: htmlEscape(unit.title),
      summary: htmlEscape(unit.summary),
      blocked_by_label: blockedByLabel,
      agents_label: agentsLabel,
      has_topology: unit.topology !== undefined,
      mermaid_graphs: mermaidGraphs,
      review_steps: unit.review_steps.map(htmlEscape),
    };
  });

  const unitCount = plan.units.length;
  const locals = {
    title: htmlEscape(plan.task_summary),
    unit_count: unitCount,
    unit_count_suffix: unitCount === 1 ? "" : "s",
    topology_count: topologyCount,
    units,
    overview_markdown_json: jsonForScript(overviewMd),
    unit_bodies_json: jsonArrayForScript(unitBodies),
    css: CSS,
    js: JS,
  };
  return eta.renderString(PLAN_TEMPLATE, locals);
}

function buildModeLabel(topology: Topology): string {
  const top: ExecutionMode = topology.execution_mode;
  const hasNestedTeam = hasNestedMode(topology.agents, "team");
  const hasNestedSubagents = hasNestedMode(topology.agents, "subagents");

  if (top === "subagents" && hasNestedTeam) return "subagents + team";
  if (top === "team" && hasNestedSubagents) return "team + subagents";
  return top;
}

function hasNestedMode(agents: Agent[], target: ExecutionMode): boolean {
  return agents.some((a) => {
    if (a.execution_mode === target) return true;
    if (a.agents) return hasNestedMode(a.agents, target);
    return false;
  });
}

function buildPhaseLabels(topology: Topology, graphCount: number): string[] {
  if (topology.execution_mode === "subagents" && graphCount > 1) {
    return Array.from({ length: graphCount }, (_, i) => `Phase ${i + 1}`);
  }
  return [];
}
